crawl-ref/source/stairs.cc

```diff
@@ -636,6 +638,42 @@ static bool _level_transition_moves_player(level_id dest,
     return dest.is_valid() != trying_to_exit;
 }
 
+level_id level_above()
+{
+    if (crawl_state.game_is_sprint())
+        return level_id();
+    // can't re-enter old zig floors; they get regenerated
+    if (player_in_branch(BRANCH_ZIGGURAT))
+        return level_id();
+    if (you.depth > 1)
+        return level_id(you.where_are_you, you.depth - 1);
+    if (!is_connected_branch(you.where_are_you))
+        return level_id(); // no rocketing out of the abyss, portals, pan...
+    const level_id entry = brentry[you.where_are_you];
+    if (entry.is_valid())
+        return entry;
+    return level_id();
+}
+
+void rise_through_ceiling()
+{
+    const level_id whither = level_above();
+    if (!whither.is_valid())
+    {
+        mpr("In a burst of heat and light, you rocket briefly upward... but you can't rise from here.");
+        return;
+    }
+
+    mpr("With a burst of heat and light, you rocket upward!");
+    untag_followers(); // XXX: is this needed?
+    floor_transition(DNGN_ALTAR_IGNIS /*hack*/, DNGN_ALTAR_IGNIS,
+                     whither, true, true, false, false);
+
+    // flavour! blow a hole through the floor
```

Comment:
for more flavour: place an Ignis altar at the departure point? Riffing on @perryprog here.

Reply:
Interesting ideas, but it feels odd to do things at the departure point (since the player won't see it until later), and not sure how much sense an altar makes flavour wise.